### DISCUSSION TITLE
Correct usage of QImage

### DISCUSSION THREAD

**Discussion by otiv-wannes-vanleemput** [KDAB/cxx-qt#871] [Q&A]
I want to visualise a Gstreamer pipeline which runs in a Rust backed. To start off, I would like to update a QImage with the current frame outputted by the pipeline. However, I'm unable to wrap a QImage in a custom rust struct.

This is my cxxqt_object.rs file:

```
#[cxx_qt::bridge]
pub mod qobject {

    unsafe extern "C++" {
        include!("cxx-qt-lib/qimage.h");
        type QImage = cxx_qt_lib::QImage;
    }

    extern "RustQt" {
        #[qobject]
        #[qml_element]
        #[qproperty(QImage, video_frame)]
        type MyObject = super::MyObjectRust;
    }

    impl cxx_qt::Constructor<(QImage,), NewArguments=(QImage,)> for MyObject {}

    unsafe extern "RustQt" {
        #[qinvokable]
        fn video_frame_from_slice(self: Pin<&mut MyObject>, image: &[u8]);
    }
}

use core::pin::Pin;
use cxx_qt_lib::{QImage};
pub struct MyObjectRust {
    video_frame: QImage,
}

impl cxx_qt::Constructor<(QImage,)> for qobject::MyObject {
    type BaseArguments = ();
    type InitializeArguments = ();
    type NewArguments = (QImage,);
    fn route_arguments(args: (QImage, )) -> (Self::NewArguments, Self::BaseArguments, Self::InitializeArguments) {
        (args, (), ())
    }
    fn new((video_frame,): (QImage,)) -> MyObjectRust {
        MyObjectRust {
            video_frame,
        }
    }
}

impl qobject::MyObject {
    pub fn video_frame_from_slice(self: Pin<&mut Self>, image: &[u8]) {
        self.video_frame = QImage::from_data(image, None).unwrap();
    }
}
```

I tried deriving the Default trait, but this resulted in a build error. Thats why I used the constructor.
However, I still get the following build error: 
```
error[E0369]: binary operation `==` cannot be applied to type `cxx_qt_lib::QImage`
 --> src/cxxqt_object.rs:3:1
  |
3 | #[cxx_qt::bridge]
  | ^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `cxx_qt::bridge` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0369`.
```

Should I fork the repository and add the PartialEq trait to QImage, or am I using it incorrectly?

**Comment by ahayzen-kdab** [comment]
Looks like we are missing `PartialEq` implementation on `QImage` which is required when it's a `Q_PROPERTY`. I'll create an issue for it to be added :-)

**Reply by LeonMatthesKDAB** [reply]
Hm, that's a case where our default generated setter may actually do the wrong thing :thinking: 
Checking image equality is potentially quite expensive if only small parts of the image change.

But I guess that's a good use-case for #42

**Comment by LeonMatthesKDAB** [comment]
@otiv-wannes-vanleemput The way you're using `QImage` in a `Q_PROPERTY` is likely not what you want.
Weirdly QML doesn't actually have a way to display a `QImage` easily.

You can tackle this issue using a `QQuickPaintedItem`. I'm unsure if that'll be fast enough for your gstreamer use-case, but it's worth a shot :)

See our training material here:
https://ferrous-systems.github.io/qt-training-2023/

Specifically the last chapter:
https://ferrous-systems.github.io/qt-training-2023/tutorial/qt-gui/index.html

**Reply by LeonMatthesKDAB** [reply]
We should of course still allow `QImage` to be used as a `Q_PROPERTY` :sweat_smile:

**Reply by otiv-wannes-vanleemput** [reply]
Thanks for the quick reply! I'll see if I'm able to get it to work with QQuickPaintedItem.
